fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/cursors/SizeStatisticsCollectorCursor.java

```diff
@@ -0,0 +1,527 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.provider.foundationdb.cursors;
+
+import com.apple.foundationdb.KeyValue;
+import com.apple.foundationdb.annotation.API;
+import com.apple.foundationdb.record.RecordCoreArgumentException;
+import com.apple.foundationdb.record.RecordCoreException;
+import com.apple.foundationdb.record.RecordCursor;
+import com.apple.foundationdb.record.RecordCursorContinuation;
+import com.apple.foundationdb.record.RecordCursorEndContinuation;
+import com.apple.foundationdb.record.RecordCursorProto;
+import com.apple.foundationdb.record.RecordCursorResult;
+import com.apple.foundationdb.record.RecordCursorVisitor;
+import com.apple.foundationdb.record.RecordMetaData;
+import com.apple.foundationdb.record.ScanProperties;
+import com.apple.foundationdb.record.metadata.Index;
+import com.apple.foundationdb.record.provider.foundationdb.FDBRecordContext;
+import com.apple.foundationdb.record.provider.foundationdb.FDBRecordStore;
+import com.apple.foundationdb.record.provider.foundationdb.KeyValueCursor;
+import com.apple.foundationdb.record.provider.foundationdb.SubspaceProvider;
+import com.apple.foundationdb.record.provider.foundationdb.SubspaceProviderBySubspace;
+import com.apple.foundationdb.subspace.Subspace;
+import com.apple.foundationdb.tuple.ByteArrayUtil2;
+import com.google.protobuf.ByteString;
+import com.google.protobuf.InvalidProtocolBufferException;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.Arrays;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+/**
+ * A cursor that collects various distribution statistics on the keys and values within a record store or within
+ * an index. The cursor performs a full scan of the subspace backing whatever it is collecting statistics on. It
+ * should therefore be run relatively sparingly, and it should not be expected to finish within a single transaction.
+ * Note that this class makes no attempt at keeping its results transactional. As a result, if this is run on an
+ * actively
+ * mutating data set, there are no guarantees that the values it returns were ever actually true for any version in the
+ * transaction history. However, as long as the data set is not too volatile, it should produce an approximate answer
+ * for the statistics it yields.
+ */
```

Comment:
I think there needs to be documentation somewhere as to what this intends to return. Based on the code, it looks like this accumulates the results across multiple runs and every result except for the last returns a `RecordCursorResult` with no result but with a continuation and a `NoNextReason`, but I don't think it's clear to the user that those are the semantics here.

Reply:
Yes, that is the behavior. I've updated the comment on the class accordingly vs. updating the onNext() method which is inheriting its Javadoc from RecordCursor. That seems to be the norm for cursors.